### PR TITLE
[GPT] enable historical signals fetch

### PR DESCRIPTION
**Pull request by GuillaumeDSM**
requires https://github.com/Drakkar-Software/OctoBot/pull/2450

### CODE REVIEW

File: Evaluator/TA/ai_evaluator/ai.py [hunk 1]
```diff
@@ -135,56 +153,75 @@ async def ohlcv_callback(self, exchange: str, exchange_id: str,
...
             return resp
         except services_errors.CreationError as err:
             raise evaluators_errors.UnavailableEvaluatorError(f"Impossible to get ChatGPT prediction: {err}") from err
+        except Exception as err:
+            print(err)
+
+    def get_version(self):
+        # later on, identify by its specs
+        # return f"{self.gpt_model}-{self.source}-{self.indicator}-{self.period}-{self.GLOBAL_VERSION}"
+        return "0.0.0"
```
Comment (Herklos) [inline Evaluator/TA/ai_evaluator/ai.py:224]: I don't understand why the version should contains all these data and not just a number (here `GLOBAL_VERISON`)

File: Evaluator/TA/ai_evaluator/ai.py [hunk 2]
```diff
@@ -100,6 +111,11 @@ def init_user_inputs(self, inputs: dict) -> None:
             self.period, inputs, min_val=1,
             title="Period: length of the indicator period."
         )
+        self.max_confidence_threshold = self.UI.user_input(
+            "max_confidence_threshold", enums.UserInputTypes.INT,
+            self.max_confidence_threshold, inputs, min_val=0, max_val=100,
+            title="Maximum confidence threshold: % confidence value starting from which to return 1 or -1."
+        )
```
Comment (Herklos) [inline Evaluator/TA/ai_evaluator/ai.py:118]: Shouldn't it be a min confidence value ?

File: Services/Services_bases/gpt_service/gpt.py [hunk 3]
```diff
@@ -63,6 +74,27 @@ async def get_chat_completion(
...
+        time_frame: str = None,
+        version: str = None,
+        candle_open_time: float = None,
+        use_stored_signals: bool = False,
+    ) -> str:
+        if use_stored_signals:
+            return self._get_signal_from_stored_signals(exchange, symbol, time_frame, version, candle_open_time)
+        if self.use_stored_signals_only():
+            return await self._fetch_signal_from_stored_signals(exchange, symbol, time_frame, version, candle_open_time)
+        return await self._get_signal_from_gpt(messages, model, max_tokens, n, stop, temperature)
```
Comment (Herklos) [inline Services/Services_bases/gpt_service/gpt.py:88]: :+1:

File: Services/Services_bases/gpt_service/gpt.py [hunk 4]
```diff
@@ -87,6 +119,111 @@ async def get_chat_completion(
...
+            for symbol in symbols
+            for time_frame in time_frames
+            if not self.has_signal_history(exchange_name, symbol, time_frame, start_timestamp, end_timestamp, version)
+        ]
+        if coros:
+            await asyncio.gather(*coros)
+
+    def _get_open_candle_timestamp(self, time_frame: commons_enums.TimeFrames, base_timestamp: float):
+        tf_seconds = commons_enums.TimeFramesMinutes[time_frame] * commons_constants.MINUTE_TO_SECONDS
+        return base_timestamp - (base_timestamp % tf_seconds)
```
Comment (Herklos) [inline Services/Services_bases/gpt_service/gpt.py:219]: Don't we have already this code somewhere else?

Comment (GuillaumeDSM) [reply Services/Services_bases/gpt_service/gpt.py:219]: added in commons

**Review by Herklos — APPROVED** [review]

**Review by Herklos — APPROVED** [review]

File: Evaluator/TA/ai_evaluator/ai.py [hunk 5]
```diff
@@ -55,7 +55,7 @@ def __init__(self, tentacles_setup_config):
         self.indicator = None
         self.source = None
         self.period = None
-        self.min_confidence_threshold = 0
+        self.min_confidence_threshold = 100
```
Comment (Herklos) [inline Evaluator/TA/ai_evaluator/ai.py:58]: This is the default value?

Comment (GuillaumeDSM) [reply Evaluator/TA/ai_evaluator/ai.py:58]: yes, we don't want to round to 1 below 100% confidence by default

Comment (Herklos) [reply Evaluator/TA/ai_evaluator/ai.py:58]: I'm not sure I understand. I haven't seen any 100% confidence value sent by gpt

**Pull request merged**